File: paint/sources/material_nodes/quantize_node.ts

function quantize_node_vector(node: ui_node_t, socket: ui_node_socket_t): string {
    let strength: string = parser_material_parse_value_input(node.inputs[0]);
    let col: string = parser_material_parse_vector_input(node.inputs[1]);
    return "(floor3(100.0 * " + strength + " * " + col + ") / (100.0 * " + strength + "))";
}

let quantize_node_def: ui_node_t = {
    id: 0,
    name: _tr("Quantize"),
    type: "QUANTIZE", // extension
    x: 0,
    y: 0,
    color: 0xff448c6d,
    inputs: [
        {
            id: 0,
            node_id: 0,
            name: _tr("Strength"),
            type: "VALUE",
            color: 0xffa1a1a1,
            default_value: f32_array_create_x(0.1),
            min: 0.0,
            max: 1.0,
            precision: 100,
            display: 0
        },
        {
            id: 0,
            node_id: 0,
            name: _tr("Color"),
            type: "RGBA",
            color: 0xffc7c729,
            default_value: f32_array_create_xyzw(0.0, 0.0, 0.0, 1.0),
            min: 0.0,
            max: 1.0,
            precision: 100,
            display: 0
        }
    ],
    outputs: [
        {
            id: 0,
            node_id: 0,
            name: _tr("Color"),
            type: "RGBA",
            color: 0xffc7c729,
            default_value: f32_array_create_xyzw(0.8, 0.8, 0.8, 1.0),
            min: 0.0,
            max: 1.0,
            precision: 100,
            display: 0
        }
    ],
    buttons: [],
    width: 0,
    flags: 0
};

File: paint/sources/material_nodes/quantize_node.c
void quantize_node_init() {
	any_array_push(nodes_material_color, quantize_node_def);
	any_map_set(parser_material_node_vectors, "QUANTIZE", quantize_node_vector);
}

char *quantize_node_vector(ui_node_t *node, ui_node_socket_t *socket) {
	char *strength = parser_material_parse_value_input(node->inputs->buffer[0], false);
	char *col      = parser_material_parse_vector_input(node->inputs->buffer[1]);
	return string_join(string_join(string_join(string_join(string_join(string_join("(floor3(100.0 * ", strength), " * "), col), ") / (100.0 * "), strength),
	                   "))");
}
